frontend/src/components/AppMap.vue

```diff
@@ -116,63 +193,169 @@ type Emits = {
 
 const emit = defineEmits<Emits>();
 
-// https://leafletjs.com/reference.html#map-option
-const mapOptions: MapOptions = {
-  // preferCanvas: true,
-  zoomSnap: 0.1,
+type Slots = {
+  "top-left": () => unknown;
+  "top-right": () => unknown;
+  "bottom-right": () => unknown;
 };
 
-// https://leafletjs.com/reference.html#layer-option
-const layerOptions: TileLayerOptions = {
-  attribution:
-    '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>',
+defineSlots<Slots>();
+
+/** elements to teleport legend template content to */
+const topLeftLegend = ref<HTMLElement>();
+const topRightLegend = ref<HTMLElement>();
+const bottomRightLegend = ref<HTMLElement>();
+
+/** element to teleport popup template content to */
+const popup = ref<HTMLElement>();
+
+/** properties of selected feature for popup */
+const popupFeature = ref<GeoJsonProperties>();
+
+/** https://leafletjs.com/reference.html#map-option */
+const mapOptions: MapOptions = {
+  zoomControl: false,
+  attributionControl: false,
 };
 
-// open street map data
-const urlTemplate = "https://tile.openstreetmap.org/{z}/{x}/{y}.png";
+/** map object */
+let map: L.Map | undefined;
+
+/** create legend panel */
+function createLegend(options: L.ControlOptions) {
+  const Control = L.Control.extend({
+    onAdd: () => document.createElement("div"),
+  });
+  const legend = new Control(options);
+  map?.addControl(legend);
+  return legend.getContainer();
+}
+
+/** set fine level of zoom snap, e.g. for tighter fit-bounds */
+function fineZoom() {
+  if (map) map.options.zoomSnap = 0.1;
+}
+
+/** set coarse level of zoom snap, to make track pad pinch zoom less frustrating */
+async function coarseZoom() {
+  /** wait for any in progress zoom to finish */
+  await sleep();
+  if (map) map.options.zoomSnap = 0.2;
+}
 
-// leaflet map object
-let map: L.Map | null = null;
+const scale = computed(() => {
```

Comment:
Unified scale computed prop that returns steps (for legend) and func for mapping value to color (for coloring features).